packages/@cdktf/hcl2cdk/lib/coerceType.ts

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) HashiCorp, Inc
+// SPDX-License-Identifier: MPL-2.0
+import template from "@babel/template";
+import * as t from "@babel/types";
+import { logger } from "@cdktf/commons";
+import { AttributeType } from "@cdktf/provider-generator";
+import { Scope } from "./types";
+import deepEqual from "deep-equal";
+
+function changeValueAccessor(
+  ast: t.MemberExpression,
+  newAccessor: string
+): t.MemberExpression {
+  const propertyIdentifier: t.Identifier = {
+    ...(ast.property as t.Identifier),
+    name: newAccessor,
+  };
+  return {
+    ...ast,
+    property: propertyIdentifier,
+  };
+}
+
+export const coerceType = (
+  scope: Scope,
+  ast: t.Expression,
+  from: AttributeType,
+  to: AttributeType | undefined
+): t.Expression => {
+  if (to === undefined) {
+    return ast;
+  }
+
+  if (deepEqual(to, from)) {
+    return ast;
+  }
+
+  const isTerraformVariableOrLocal =
+    ast.type === "MemberExpression" &&
+    ast.property.type === "Identifier" &&
+    ast.property.name === "value" &&
+    ast.object.type === "Identifier" &&
+    Object.values(scope.variables).some(
+      (knownVars) =>
+        knownVars.variableName === (ast.object as t.Identifier).name
+    );
```

Comment:
Would this match e.g. `someResource.value`? There are [multiple](https://github.com/search?q=repo%3Acdktf%2Fcdktf-provider-aws%20value&type=code) of those in e.g. the AWS TF provider bindings 😬

Reply:
Yes good point, scope.variables holds all TS-variable names, therefore also resources that are referenced. I added a new property to the map now indicating if the variable is varOrLocal, so that this can not lead to errors

Reply:
I added a check for the resource type to find variables and locals specifically